sdk/packages/core/src/contracts/watchDidFill.ts

```diff
@@ -18,16 +19,60 @@ export function watchDidFill({
   outboxAddress,
   orderId,
   pollingInterval,
-  onLogs,
+  onFill,
   onError,
-}: WatchDidFillParams): WatchDidFillReturn {
-  return watchContractEvent(client, {
-    address: outboxAddress,
-    eventName: 'Filled',
-    abi: outboxABI,
-    args: { orderId },
-    onLogs,
-    onError,
+}: WatchDidFillParams) {
+  return watchBlocks(client, {
+    onBlock: async (block: Block) => {
+      try {
+        // default block tag is 'latest', hash is only undefined for pending blocks
+        invariant(!!block.hash, 'Block hash is undefined')
+
+        const log = await getFilledLog({
+          client,
+          outboxAddress,
+          orderId,
+          blockHash: block.hash,
+        })
+
+        if (log) onFill(log.transactionHash)
+      } catch (error) {
+        onError?.(error as Error)
+      }
+    },
     pollingInterval,
+    onError,
+    emitMissed: true,
+  })
+}
+
+async function getFilledLog({
+  client,
+  outboxAddress,
+  orderId,
+  blockHash,
+}: Pick<WatchDidFillParams, 'client' | 'outboxAddress' | 'orderId'> & {
+  blockHash: Hex
+}) {
+  const logs = await getLogs(client, {
+    address: outboxAddress,
+    event: getFilledEventAbi(),
+    args: {
+      orderId,
+    },
```

Comment:
This is fine to start. Though technically this report an order as Filled, even if our solver filled the wrong call.
More complete would be to use the [`fillHash`](https://github.com/omni-network/omni/blob/f7703137d1374af01416224beefff3854edecdc0/contracts/solve/src/SolverNetOutbox.sol#L308) i args as well

Though can do later, orderId being insufficient is v unlikely at this point